scaluq/gate/gate_matrix.hpp

```diff
@@ -99,8 +180,619 @@ class TwoQubitMatrixGateImpl : public TwoQubitGateBase {
+
+    const UINT target_qubit_index_count = target_qubit_index_list.size();
+    const UINT matrix_dim = 1ULL << target_qubit_index_count;
+    const std::vector<UINT> matrix_mask_list =
+        create_matrix_mask_list(target_qubit_index_list, target_qubit_index_count);
+
+    const UINT loop_dim = state.dim() >> target_qubit_index_count;
+    auto mask_array_view = convert_host_vector_to_device_view(mask_array);
+    auto matrix_mask_list_view = convert_host_vector_to_device_view(matrix_mask_list);
+
+    Kokkos::parallel_for(
+        Kokkos::TeamPolicy<>(loop_dim, Kokkos::AUTO()),
+        KOKKOS_LAMBDA(const Kokkos::TeamPolicy<>::member_type& team) {
+            auto buffer = Kokkos::View<Complex*>("buffer", matrix_dim);
+            UINT basis_0 = team.league_rank();
+            for (UINT cursor = 0; cursor < target_qubit_index_count; ++cursor) {
+                basis_0 = (basis_0 & mask_array_view[cursor]) +
+                          ((basis_0 & (~mask_array_view[cursor])) << 1);
+            }
+
+            Kokkos::parallel_for(Kokkos::TeamThreadRange(team, matrix_dim), [&](UINT y) {
+                buffer[y] = 0;
+                Complex sum = 0;
+                Kokkos::parallel_reduce(
+                    Kokkos::ThreadVectorRange(team, matrix_dim),
+                    [&](UINT x, Complex& inner_sum) {
+                        inner_sum += matrix(y, x) * state._raw[basis_0 ^ matrix_mask_list[x]];
+                    },
+                    sum);
+                buffer[y] = sum;
+            });
+            team.team_barrier();
+
+            Kokkos::parallel_for(Kokkos::TeamThreadRange(team, matrix_dim), [=](const UINT y) {
+                state._raw[basis_0 ^ matrix_mask_list[y]] = buffer[y];
+            });
+        });
+    Kokkos::fence();
+}
+
+inline void multi_qubit_dense_matrix_gate(const std::vector<UINT>& target_qubit_index_list,
+                                          const Matrix& matrix,
+                                          StateVector& state_vector) {
+    UINT target_qubit_index_count = target_qubit_index_list.size();
+    if (target_qubit_index_count == 1) {
+        single_qubit_dense_matrix_gate_view(target_qubit_index_list[0], matrix, state_vector);
+        return;
+    } else if (target_qubit_index_count == 2) {
+        double_qubit_dense_matrix_gate(
+            target_qubit_index_list[0], target_qubit_index_list[1], matrix, state_vector);
+        return;
+    } else {
+        multi_qubit_dense_matrix_gate_parallel(target_qubit_index_list, matrix, state_vector);
+        return;
+    }
+}
 }  // namespace internal
 
 using OneQubitMatrixGate = internal::GatePtr<internal::OneQubitMatrixGateImpl>;
 using TwoQubitMatrixGate = internal::GatePtr<internal::TwoQubitMatrixGateImpl>;
+using SparseMatrixGate = internal::GatePtr<internal::SparseMatrixGateImpl>;
+using DensityMatrixGate = internal::GatePtr<internal::DenseMatrixGateImpl>;
```

Comment:
```suggestion
using DenseMatrixGate = internal::GatePtr<internal::DenseMatrixGateImpl>;
```
DensityMatrixは別物の意味になります

Reply:
修正漏れがありました...直しておきました